dhall-docs/src/Dhall/Docs/Core.hs

```diff
@@ -381,7 +382,7 @@ generateDocs
     -> IO ()
 generateDocs inputDir outLink packageName = do
     (_, absFiles) <- Path.IO.listDirRecur inputDir
-    contents <- mapM (Text.IO.readFile . Path.fromAbsFile) absFiles
+    contents <- mapM (Text.IO.readFile . Path.fromAbsFile) (filter hasDhallExtension absFiles)
```

Comment:
I'd prefer to keep the filtering in just a single place (i.e. `getAllDhallFiles`)

I've tried your patch locally and it fails with binary files that have non-utf8 characters even with files with ".dhall" extension.

I've faced this issue before, and the solution was to:

1. Read files as `ByteString` i.e. using `ByteString.readfile`
2. Encode as `Text` utf-8 i.e. using `Data.Text.Encoding.decodeUtf8'`
3. Filter only the files that don't fail to decode without filtering again by .dhall extension. As I said above, the `.dhall` filtering is done by `getAllDhallfiles`.

Reply:
If you don't have time for the patch I can work on it later :smile: I'm currently working on #1908 and it will take some time.

Reply:
I don't understand, `getAllDhallFiles` is not called by `generateDocs` which seems to be failing when running dhall-docs on binary files. `generateDocs` is already using `Text.IO.readFile`, perhaps we could simply ignore the files that raised the invalid byte sequence exception?

Reply:
`generateDocs` calls `generateDocsPure` which calls `getAllDhallFiles`.

That fix fails even with binary files with `.dhall` extension. The steps I pointed you in my comment are how I fixed this error the time I encountered it (weeks ago) and what probably would work to fix the issue.

`Data.Text.IO.readFile` fails reading binary files, `Data.ByteString.readFile` doesn't. You can read the file using the latest and try to decode it as utf8 using `Data.Text.Encoding.decodeUtf8'` to also do the filtering.

Reply:
I see, thank you very much for the detailed explanation.